utils/linear_probes.py

```diff
@@ -0,0 +1,298 @@
+        logits, activations = model.run_with_cache(d)
+        for act_str in activations.keys():
+            if 'hook_resid_post' in act_str: # only take activations after residual connection
+                acts_dict[act_str].append(copy(activations[act_str]).detach().cpu().numpy())
+    return acts_dict
+
+
+def train_linear_probe(x1, x2):
+    # concatenate the two datasets
+    x = rearrange([x1, x2], 'x n d -> (x n) d')
+    # labels: zero for data1, one for data2
+    y = rearrange([np.zeros(len(x1)), np.ones(len(x2))], 'x n -> (x n)')
+
+    x, y = shuffle(x, y, random_state=0)
+
+    # train a linear probe with l2 regularization and 5 fold cross validation
+    clf = LogisticRegression(random_state=0, max_iter=1000, penalty='l2', C=1.0)
+    scores = cross_val_score(clf, x, y, cv=5, scoring='accuracy', n_jobs=4)
+    return scores
+
+
+def leave_unique_q_type(data: List[str], model: HookedTransformer, q_type:str='born', filter_var_len=3) -> List[str]:
+    """
+    q_type is a string that must be contained in the question
+    model is needed for tokenization
+    filter_var_len is the length (in tokens) of the variables we keep
+    """
+    out = []
+    for d in data:
+        assert d.count('<|') == 1, f'{d} is a definition'  # ensure "<|" occurs only once in the string     
+        var = d.split('<|')[1].split('|>')[0]  # variable is surrounded by <| and |>
+        if q_type in d and len(model.tokenizer.encode(var)) == filter_var_len:
+            out.append(d)
+    assert len(out) > 0, f'no data for q_type {q_type} and filter_var_len {filter_var_len}'
+    # ensure all questions have the same tokenized length
+    assert all([len(model.tokenizer.encode(d)) == len(model.tokenizer.encode(out[0])) for d in out])
+    return out
+    
+    
+def run_q_type(model, data1, data2, q_type='born', filter_var_len=3, device='cuda'):
+    """Train linear probes to distinguish between data1 and data2 based on activations."""
+    if type(model) == str:
+        model = HookedTransformer.from_pretrained(model, device=device)
+    
+    # We need to ensure that the same variable cannot be in the train and in the test set
+    # and that all questions have the same tokenized length. Simplest solution: take only one type of question
+    data1 = leave_unique_q_type(data1, model, q_type, filter_var_len)
+    data2 = leave_unique_q_type(data2, model, q_type, filter_var_len)
+    assert len(model.tokenizer.encode(data1[0])) == len(model.tokenizer.encode(data2[0]))
+    
+    # make sure data1 and data2 have the same length so that baseline linear probe accuracy is 0.5
+    minlen = min(len(data1), len(data2))
+    data1, data2 = data1[:minlen], data2[:minlen]
+    print(f'data lengths: {len(data1)}, {len(data2)}')
+    
+    acts_data1 = get_activations(model, data1)
+    acts_data2 = get_activations(model, data2)
+    
+    score_grid = []
+    # iterate over tokens
+    for i in range(len(model.tokenizer.encode(data1[0]))):
```

Comment:
is this score grid made only for the first example from data1 and data2?

Reply:
No, see the get_activations function

Reply:
`len(model.tokenizer.encode(data1[0]))` is just a way to get the number of tokens (can get it from acts_data too)

Reply:
I did a while ago!